include/picongpu/plugins/Emittance.hpp

```diff
@@ -0,0 +1,1051 @@
+         * gSumMom2, gSumPos2, gSumMomPos2, gCount_e
+         */
+        void notify( uint32_t currentStep )
+        {
+            // call the method that calls the plugin kernel
+            calculateCalcEmittance < CORE + BORDER > ( currentStep );
+        }
+
+
+        void restart(
+            uint32_t restartStep,
+            std::string const & restartDirectory
+        )
+        {
+            if( !writeToFile )
+                return;
+
+            writeToFile = restoreTxtFile(
+                outFile,
+                filename,
+                restartStep,
+                restartDirectory
+            );
+        }
+
+        void checkpoint(
+            uint32_t currentStep,
+            std::string const & checkpointDirectory
+        )
+        {
+            if( !writeToFile )
+                return;
+
+            checkpointTxtFile(
+                outFile,
+                filename,
+                currentStep,
+                checkpointDirectory
+            );
+        }
+
+    private:
+        //! method to call analysis and plugin-kernel calls
+        template< uint32_t AREA >
+        void calculateCalcEmittance( uint32_t currentStep )
+        {
+            DataConnector &dc = Environment<>::get( ).DataConnector( );
+
+            // use data connector to get particle data
+            auto particles = dc.get< ParticlesType >(
+                ParticlesType::FrameType::getName( ),
+                true
+            );
+
+            // initialize global gSum... with zero
+            gSumMom2->getDeviceBuffer( ).setValue( 0.0 );
+            gSumPos2->getDeviceBuffer( ).setValue( 0.0 );
+            gSumMomPos2->getDeviceBuffer( ).setValue( 0.0 );
+            gCount_e->getDeviceBuffer( ).setValue( 0.0 );
+
+            constexpr uint32_t numWorkers = pmacc::traits::GetNumWorkers<
+                pmacc::math::CT::volume< SuperCellSize >::type::value
+            >::value;
+
+            AreaMapping<
+                AREA,
+                MappingDesc
+            > mapper( *m_cellDescription );
+
+            auto kernel = PMACC_KERNEL( KernelCalcEmittance< numWorkers >{ } )(
+                mapper.getGridDim( ),
+                numWorkers
+            );
+
+            // Some funny things that make it possible for the kernel
+            // to calculate the absolute position of the particles
+            DataSpace<simDim> localSize(m_cellDescription->getGridLayout().getDataSpaceWithoutGuarding());
+            const SubGrid<simDim>& subGrid = Environment<simDim>::get().SubGrid();
+            const int subGridY = subGrid.getGlobalDomain().size.y();
+            auto movingWindow=MovingWindow::getInstance().getWindow(currentStep);
+            DataSpace<simDim> globalOffset(subGrid.getLocalDomain().offset);
+            auto globalWindowOffset = movingWindow.globalDimensions.offset.y();
+            globalOffset.y() -= globalWindowOffset;
```

Comment:
probably outdated, needs to be removed

Reply:
confirmed, needs to go

Reply:
not sure what `globalOffset` in this context should be. I assume it is the offset from the window begin to the gpu/device. Is this correct?